solutions/AC Matching.mdx

```diff
@@ -0,0 +1,58 @@
+---
+id: ac-matching
+source: AtCoder DP Contest
+title: Matching
+author: Neo Wang
+---
+
+## Explanation
+
+If we define $\texttt{dp}[S]$ to be the amount of matchings for females in set
+$S$ to the first $|S|$ males, this problem boils down to the following
+formulation:
+
+$$
+\texttt{dp}[S] = \sum \texttt{dp}[S\backslash x]\: \text{if}\:  \texttt{compatible}[|S|][x]
+$$
+
+In English, this is the equivalent of
+
+> The amount of matchings in a subset $S$ to include a certain female $x$ is
+> equivalent to the sum of all the matchings without female $x$ such that the
+> addition of female $x$ is compatible with the $|S|$-th male.
+
+Our base case is the empty set, which has a value of $1$ (the empty set can be
+considered as a single matching involving zero pairs).
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N\cdot 2^N)$
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+bool compat[21][21];
+int dp[1 << 21];
+
+int N;
+
+int main() {
+	cin.tie(0)->sync_with_stdio(0);
+	cin >> N;
+	for(int i = 0; i < N; i++) for(int j = 0; j < N; j++) cin >> compat[i][j];
+
+	dp[0] = 1;
+
+	for(int s = 0; s < (1<<N); s++) {
+		int v = __builtin_popcount(s);
+		for(int x = 0; x < N; x++) {
+			if((s & (1 << x)) || !compat[v][x]) continue;
```

Comment:
`(s & (1 << x))`

why is there space for bitwise and but not bitwise or? Same question regarding the bit shifts above and below